lib/spack/spack/architecture.py

```diff
@@ -56,49 +56,133 @@
 from llnl.util.lang import memoized, list_modules, key_ordering
 
 import spack.compiler
 import spack.paths
 import spack.error as serr
+import spack.version
 from spack.util.naming import mod_to_class
 from spack.util.spack_yaml import syaml_dict
 
 
 class NoPlatformError(serr.SpackError):
-
     def __init__(self):
         super(NoPlatformError, self).__init__(
             "Could not determine a platform for this machine.")
 
 
-@key_ordering
-class Target(object):
-    """ Target is the processor of the host machine.
-        The host machine may have different front-end and back-end targets,
-        especially if it is a Cray machine. The target will have a name and
-        also the module_name (e.g craype-compiler). Targets will also
-        recognize which platform they came from using the set_platform method.
-        Targets will have compiler finding strategies
+def _ensure_other_is_target(method):
+    """Decorator to be used in dunder methods taking a single argument to
+    ensure that the argument is an instance of ``Target`` too.
     """
+    @functools.wraps(method)
+    def _impl(self, other):
+        if isinstance(other, six.string_types):
+            other = Target(other)
+
+        if not isinstance(other, Target):
+            return NotImplemented
+
+        return method(self, other)
 
+    return _impl
+
+
+class Target(object):
     def __init__(self, name, module_name=None):
-        self.name = name  # case of cray "ivybridge" but if it's x86_64
-        self.module_name = module_name  # craype-ivybridge
+        """Target models the processor of the current architecture.
+
+        Args:
+            name (str or MicroArchitecture):micro-architecture of the
+                target
+            module_name (str): optional module name to get access to the
+                current target. This is typically used on machines
+                like Cray (e.g. craype-compiler)
+        """
+        if not isinstance(name, cpu.MicroArchitecture):
+            name = cpu.targets.get(
+                name, cpu.generic_microarchitecture(name)
+            )
+        self.micro_architecture = name
+        self.module_name = module_name
 
-    # Sets only the platform name to avoid recursiveness
+    @property
+    def name(self):
+        return self.micro_architecture.name
 
-    def _cmp_key(self):
-        return (self.name, self.module_name)
+    @_ensure_other_is_target
+    def __eq__(self, other):
+        return self.micro_architecture == other.micro_architecture and \
+            self.module_name == other.module_name
+
+    def __ne__(self, other):
+        # This method is necessary as long as we support Python 2. In Python 3
+        # __ne__ defaults to the implementation below
+        return not self == other
+
+    @_ensure_other_is_target
+    def __lt__(self, other):
+        # This is needed to sort deterministically specs in a list.
+        # It doesn't implement a total ordering semantic.
```

Comment:
This probably needs a TODO to think about what to do long-term.  I think it's convenient to be able to say `spec.architecture.target < other.architecture.target` but I think this implementation is a safer choice for now.

Reply:
Added a TODO